kedro/io/data_catalog.py

```diff
@@ -567,16 +642,47 @@ def list(self, regex_search: str | None = None) -> list[str]:
             ) from exc
         return [dset_name for dset_name in self._data_sets if pattern.search(dset_name)]
 
+    def exists_in_catalog_config(self, dataset_name: str) -> bool:
+        """Check if a dataset exists in the catalog as an exact match or if it matches a pattern."""
+        if (
+            dataset_name in self._data_sets
+            or dataset_name in self._pattern_matches_cache
+        ):
+            return True
+        matched_pattern = self.match_name_against_patterns(dataset_name)
+        if matched_pattern:
+            # cache the "dataset_name -> pattern" match
+            self._pattern_matches_cache[dataset_name] = matched_pattern
+            return True
+        return False
+
+    def match_name_against_patterns(self, dataset_name: str) -> str | None:
+        """Match a dataset name against existing patterns"""
+        # Loop through all dataset patterns and check if the given dataset name has a match.
+        for pattern in self._sorted_dataset_patterns:
+            result = parse(pattern, dataset_name)
+            if result:
+                return pattern
+        return None
+
     def shallow_copy(self) -> DataCatalog:
         """Returns a shallow copy of the current object.
 
         Returns:
             Copy of the current object.
         """
-        return DataCatalog(data_sets=self._data_sets, layers=self.layers)
+        return DataCatalog(
+            data_sets=self._data_sets,
+            layers=self.layers,
+            dataset_patterns=self.dataset_patterns,
+        )
 
     def __eq__(self, other):
```

Comment:
Maybe a good exercise to use dataclass or attr to remove these boilerplate later